### PR TITLE
[Explicit Module Builds] Support explicit builds of packages witn inter-dependent Swift targets

### PR DESCRIPTION
**Pull request by artemcm**
Building a Swift module in Explicit Module Build mode requires passing all of its module dependencies as explicit arguments to the build command. Thus, building a SwiftPM package with multiple inter-dependent targets requires that each target’s build job must have its target dependencies’ modules passed into it as explicit module dependencies.

When planning a build that involves multiple Swift modules (targets), because planning takes place before any of the targets are built, the dependency scanning action will not be able to discover such modules. In such cases, we must provide the driver with information about all such external dependencies, including the path to where their compiled .swiftmodule will be located, once built, and a full inter-module dependency graph for each such dependency. Only then will the driver be able to create explicit-module build jobs for targets with dependencies.

This PR adds explicit package build handling by planning builds for all targets in a topological order and directly passing information about target's dependencies (module path and dependency graph) into the driver planning action. 

This change works in tandem with the corresponding Dependency Scanner change in:
https://github.com/apple/swift/pull/33032
And a corresponding swift-driver change:
https://github.com/apple/swift-driver/pull/181

Both of which will have to be submitted first. 
cc @nkcsgexi

### CODE REVIEW

**Comment by artemcm** [comment]
@swift-ci please smoke test

**Comment by artemcm** [comment]
@DougGregor, this mostly refactors the code that involves invoking the integrated driver during build planning. Could you please take a look?

File: Sources/Build/ManifestBuilder.swift [hunk 1]
```diff
@@ -224,8 +230,19 @@ extension LLBuildManifestBuilder {
                                     fileSystem: target.fs,
                                     executor: executor)
             let jobs = try driver.planBuild()
-            let resolver = try ArgsResolver(fileSystem: target.fs)
+            addSwiftDriverJobs(for: target, jobs: jobs, inputs: inputs,
+                               isMainModule: { driver.isExplicitMainModuleJob(job: $0)})
+         } catch {
+             fatalError("\(error)")
```
Comment (DougGregor) [inline Sources/Build/ManifestBuilder.swift:236]: Why the fatal error here? Can't we `throw` out of this method?

File: Sources/Build/ManifestBuilder.swift [hunk 2]
```diff
@@ -268,12 +285,147 @@ extension LLBuildManifestBuilder {
                         args: arguments
                     )
                 }
-             }
+            }
+        } catch {
+            fatalError("\(error)")
```
Comment (DougGregor) [inline Sources/Build/ManifestBuilder.swift:290]: Same question: seems like we should be able to `throw` out of here? Fatal errors should be due to logic errors in the code, only.

File: Tests/BuildTests/BuildPlanTests.swift [hunk 3]
```diff
@@ -163,6 +167,90 @@ final class BuildPlanTests: XCTestCase {
       #endif
     }
 
+    func testExplicitSwiftPackageBuild() throws {
+        #if os(macOS)
```
Comment (DougGregor) [inline Tests/BuildTests/BuildPlanTests.swift:171]: Why macOS-only?

File: Sources/Build/ManifestBuilder.swift [hunk 4]
```diff
@@ -268,12 +285,147 @@ extension LLBuildManifestBuilder {
...
+        let cmdOutputs = objectNodes + [moduleNode]
+
+        // Commands.
+        addExplicitBuildSwiftCmds(description, inputs: inputs,
+                                  objectNodes: objectNodes,
+                                  moduleNode: moduleNode,
+                                  targetDepGraphMap: &targetDepGraphMap)
+
+        addTargetCmd(description, cmdOutputs: cmdOutputs)
+        addModuleWrapCmd(description)
```
Comment (DougGregor) [inline Sources/Build/ManifestBuilder.swift:362]: Probably unrelated to your PR, but I thought the driver took care of the module-wrap command?

File: Sources/Build/ManifestBuilder.swift [hunk 5]
```diff
@@ -268,12 +285,147 @@ extension LLBuildManifestBuilder {
...
+            }
+            targetDepGraphMap[targetDescription.target] = dependencyGraph
+            addSwiftDriverJobs(for: targetDescription, jobs: jobs, inputs: inputs,
+                               isMainModule: { driver.isExplicitMainModuleJob(job: $0)})
          } catch {
              fatalError("\(error)")
          }
     }
 
+    private func collectTargetDependencyInfos(for target: ResolvedTarget,
```
Comment (DougGregor) [inline Sources/Build/ManifestBuilder.swift:405]: Could you comment this function? It'd be worth pointing out, for example, that this is the transitive list of all dependencies. Giving the result type a `typealias` with a meaningful name would help a lot. I see, for example, that the `AbsolutePath` is the `.swiftmodule` file? Might we need to consider that ` .swiftinterface` would end up here?

**Comment by artemcm** [comment]
@swift-ci please smoke test

**Comment by artemcm** [comment]
@swift-ci please smoke test

**Comment by artemcm** [comment]
@swift-ci please smoke test

**Comment by artemcm** [comment]
It seems that the `self-hosted` CI jobs are not using the latest toolchain snapshot, they are using the machine's installed Xcode toolchain. 

@neonichu, @aciidb0mb3r, do we specifically want this to test against some stable toolchain or would it be okay to switch the `self-hosted` CI to use the latest development snapshot? For example, this PR adds a test that needs to invoke the integrated driver, which nowadays is very tightly coupled to some recent `swift` changes.

**Comment by aciidb0mb3r** [comment]
My hope with the self hosted bot was to enable development with latest Xcode (including betas). Maybe it’s possible to make the test conditional somehow?

**Comment by artemcm** [comment]
@swift-ci please smoke test

**Pull request merged**